impl/src/test/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindTestBase.java

```diff
@@ -29,113 +25,223 @@
 
 import org.ehcache.Cache;
 import org.ehcache.CacheManager;
+import org.ehcache.config.CacheConfiguration;
+import org.ehcache.config.CacheConfigurationBuilder;
 import org.ehcache.exceptions.BulkCacheWritingException;
+import org.ehcache.expiry.Duration;
+import org.ehcache.expiry.Expirations;
+import org.ehcache.spi.loaderwriter.CacheLoaderWriter;
+import org.ehcache.spi.loaderwriter.CacheLoaderWriterProvider;
 import org.junit.Test;
 
+import static org.ehcache.CacheManagerBuilder.newCacheManagerBuilder;
+import static org.ehcache.config.writebehind.WriteBehindConfigurationBuilder.newWriteBehindConfiguration;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.collection.IsIterableContainingInOrder.contains;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Matchers.anyObject;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
 /**
  * @author Abhilash
  *
  */
 public abstract class AbstractWriteBehindTestBase {
   
-  protected CacheManager cacheManager ;
-  protected Cache<String, String> testCache;
-  protected final WriteBehindTestLoaderWriter<String, String> loaderWriter = new WriteBehindTestLoaderWriter<String, String>();
+  protected abstract CacheConfigurationBuilder<Object, Object> configurationBuilder();
+
+  @Test
+  public void testWriteOrdering() throws Exception {
+    WriteBehindTestLoaderWriter<String, String> loaderWriter = new WriteBehindTestLoaderWriter<String, String>();
+    CacheLoaderWriterProvider cacheLoaderWriterProvider = mock(CacheLoaderWriterProvider.class);
+    when(cacheLoaderWriterProvider.createCacheLoaderWriter(anyString(), (CacheConfiguration<String, String>)anyObject())).thenReturn((CacheLoaderWriter)loaderWriter);
+    
+    CacheManager cacheManager = newCacheManagerBuilder().using(cacheLoaderWriterProvider).build(true);
+    try {
+      Cache<String, String> testCache = cacheManager.createCache("testWriteOrdering", configurationBuilder()
+          .add(newWriteBehindConfiguration().queueSize(10).batchSize(10).delay(1, 2).build())
+          .buildConfig(String.class, String.class));
+
+      CountDownLatch countDownLatch = new CountDownLatch(8);
+
+      loaderWriter.setLatch(countDownLatch);
+      
+      testCache.remove("key");
+      testCache.put("key", "value1");
+      testCache.remove("key");
+      testCache.put("key", "value2");
+      testCache.remove("key");
+      testCache.put("key", "value3");
+      testCache.remove("key");
+      testCache.put("key", "value4");
```

Comment:
Why do we do 8 operations when batch is configured at 10? This forces the whole system to wait for the max delay.